src/components/recordedit/form-row.tsx

```diff
@@ -0,0 +1,285 @@
+import SelectAllRow from '@isrd-isi-edu/chaise/src/components/recordedit/select-all-row';
+
+// hooks
+import { useEffect, useLayoutEffect, useRef, useState } from 'react';
+import useRecordedit from '@isrd-isi-edu/chaise/src/hooks/recordedit';
+
+// models
+import { appModes, SELECT_ALL_INPUT_FORM_VALUE } from '@isrd-isi-edu/chaise/src/models/recordedit';
+
+// utils
+import { getDisabledInputValue } from '@isrd-isi-edu/chaise/src/utils/input-utils';
+import ResizeSensor from 'css-element-queries/src/ResizeSensor';
+import { isObjectAndKeyDefined } from '@isrd-isi-edu/chaise/src/utils/type-utils';
+import { makeSafeIdAttr } from '@isrd-isi-edu/chaise/src/utils/string-utils';
+
+type FormRowProps = {
+    columnModelIndex: number;
+    removeFormIndex?: number;
+    removeClicked?: boolean;
+    setRemoveClicked?: any;
+};
+const FormRow = ({
+    columnModelIndex,
+    removeFormIndex,
+    removeClicked,
+    setRemoveClicked,
+}: FormRowProps): JSX.Element => {
+    const {
+        forms,
+        appMode,
+        reference,
+        columnModels,
+        tuples,
+        activeSelectAll,
+        canUpdateValues,
+        columnPermissionErrors,
+        foreignKeyData,
+        waitingForForeignKeyData,
+        getRecordeditLogStack,
+        getRecordeditLogAction,
+    } = useRecordedit();
+
+    // This state variable is to set the form as active when its selected
+    const [activeForms, setActiveForm] = useState<number[]>([]);
+    /**
+     * which columns should show the permission error.
+     * if a user cannot edit a column in one of the rows, we cannot allow them
+     * to edit that column in other rows.
+     */
+    const [showPermissionError, setShowPermissionError] = useState<{
+        [key: string]: boolean;
+    }>({});
+
+    /**
+     * reset the state of showing permission errors whenever the errors changed
+     */
+    useEffect(() => {
+        setShowPermissionError({});
+    }, [columnPermissionErrors]);
+
+    const container = useRef<HTMLDivElement>(null);
+    const cm = columnModels[columnModelIndex];
+
+    /**
+     * make sure the column names (key-column.tsx) have the same height as FormRow
+     */
+    useLayoutEffect(() => {
+        if (!container || !container.current) return;
+
+        let cachedHeight = -1;
+        const sensor = new ResizeSensor(
+            container.current as Element,
+            (dimension) => {
+                const newHeight = container.current?.getBoundingClientRect().height;
+                if (newHeight === undefined || newHeight === cachedHeight || !container.current) return;
+                cachedHeight = newHeight;
+                const header = document.querySelector<HTMLElement>(`.entity-key.entity-key-${columnModelIndex}`);
+                if (header) {
+                    header.style.height = `${cachedHeight}px`;
+                }
+            }
+        );
+
+        return () => {
+            sensor.detach();
+        };
+    }, []);
+    /*
```

Comment:
Please use proper style of comments...

Reply:
You should also explain that we're doing it this way because the `removeClicked` comes from the parent and this is basically how we're communicating from parent to this child component.